rclcpp/src/rclcpp/executor.cpp

```diff
@@ -234,13 +234,18 @@ Executor::spin_some(std::chrono::nanoseconds max_duration)
     throw std::runtime_error("spin_some() called while already spinning");
   }
   RCLCPP_SCOPE_EXIT(this->spinning.store(false); );
+  // during the first run only, the process is allowed to check for new work
+  bool allow_wait = true;
   while (spinning.load() && max_duration_not_elapsed()) {
     AnyExecutable any_exec;
-    if (get_next_executable(any_exec, std::chrono::milliseconds::zero())) {
+    if (get_next_executable(any_exec, std::chrono::milliseconds::zero(), allow_wait)) {
```

Comment:
We're passing `std::chrono::milliseconds::zero()`, so this is never waiting at all.

Reply:
The reason for the added flag is actually beyond the amount of time that's used for waiting. Even with a zero wait duration, the `wait_for_work` method is still called, and it does several things beyond simply waiting for the timeout to occur. Most notably, it calls `add_handles_to_wait_set`, which will requery the various sources for work. This is one of the two ways that "extra" work can be considered during a run of `spin_some` (the other being the timers). By adding a flag that prevents `wait_for_work` from being called repeatedly, we can avoid this rebuilding of the work queues.

Perhaps it would be better if the flag were renamed, perhaps to `allow_new_work`?